### PR TITLE
Tweak configuration of JsonSerializerOptions

### LINKED ISSUE #49849: Investigate thread safety of GetReadOnlyTypeInfo and SystemTextJsonOutputFormatter
It came up in API review for [MakeReadOnly(bool populateMissingResolver)](https://github.com/dotnet/runtime/issues/89934) that the pattern of conditionally setting `JsonSerializerOptions.TypeInfoResolver` before calling `MakeReadOnly` is not thread safe and can lead to InvalidOperationExceptions like those in https://github.com/dotnet/runtime/issues/89830 when there's a race.

At first blush, [GetReadOnlyTypeInfo](https://github.com/dotnet/aspnetcore/blob/f1cfa8ea0bd80096894dd210e754db4dafa83b62/src/Shared/Json/JsonSerializerExtensions.cs#L19-L25) and [the SystemTextJsonOutputFormatter ctor](https://github.com/dotnet/aspnetcore/blob/f1cfa8ea0bd80096894dd210e754db4dafa83b62/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs#L26-L28) appear to have the same thread safety issue. Of course, if we can guarantee these never run concurrently, there might not be an issue. That could explain why our tests haven't caught this, but I think it's more likely that we just don't have any functional tests where `TypeInfoResolver` would be null by the time it gets to this code.

https://github.com/dotnet/aspnetcore/blob/f1cfa8ea0bd80096894dd210e754db4dafa83b62/src/Shared/Json/JsonSerializerExtensions.cs#L19-L25

https://github.com/dotnet/aspnetcore/blob/f1cfa8ea0bd80096894dd210e754db4dafa83b62/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs#L26-L28

It appears we won't be able to use the newly approved thread-safe `MakeReadOnly(bool populateMissingResolver)` API because we try to set an empty rather than the default reflection-based `TypeInfoResolver`. I'm not sure why that is. Is it an attempt to reduce the need for suppressions when we're doing serialization that is not statically verifiable to be linker safe?

I get that it's probably unusual to get a null `TypeInfoResolver` unless reflection is really not supported given who `JsonOptions.DefaultSerializerOptions` gets configured, but it still seems nicer to at least try to fallback to reflection rather than just fail to serialize or deserialize if there's even a small chance that it might work. Wouldn't it effectively just change the exception that's thrown if the reflection fallback fails? Probably to something with a clearer error message?

@eerhardt @eiriktsarpalis @captainsafia

### LINKED ISSUE #49855: RDG throws at startup when combined with PublishAot and AuthorizationMiddleware
`RequestDelegateGenerator` throws at startup when the `AuthorizationMiddleware` constructor resolves its `AuthorizationPolicyCache` and `PublishAot` is enabled. The throwing call to `GetTypeInfo` is in the generated code's `RequestDelegateFactoryFunc createRequestDelegate` in lines generated for the filter pipeline to handle possible object returns with `objectJsonTypeInfo`.

https://github.com/dotnet/aspnetcore/blob/dbd6a51154e0fbab890364372a1db6c1f082a8fd/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs#L9-L14

I'm not exactly sure why adding the `AuthorizationMiddleware` triggers it, but here's the repro app:

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddAuthorization();
var app = builder.Build();

app.MapGet("/", () => "Hello, World!");

app.Run();
```

<details><summary>Hand here's the unhandled startup exception:</summary>

```
Unhandled exception. System.NotSupportedException: JsonTypeInfo metadata for type 'System.Object' was not provided by TypeInfoResolver of type '<null>'. If using source generation, ensure that all root types passed to the serializer have been annotated with 'JsonSerializableAttribute', along with any types that might be serialized polymorphically.
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException_NoMetadataForType(Type type, IJsonTypeInfoResolver resolver)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoInternal(Type type, Boolean ensureConfigured, Nullable`1 ensureNotNull, Boolean resolveIfMutable, Boolean fallBackToNearestAncestorType)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfo(Type type)
   at Microsoft.AspNetCore.Http.Generated.<GeneratedRouteBuilderExtensions_g>F69328E0708B4B584C5AACA22FE2C51A1CF192D6622828F613FC57C583CA77B63__GeneratedRouteBuilderExtensionsCore.<>c.<MapGet0>b__1_1(Delegate del, RequestDelegateFactoryOptions options, RequestDelegateMetadataResult inferredMetadataResult) in D:\dev\halter73\JwtBearerAot\Microsoft.AspNetCore.Http.RequestDelegateGenerator\Microsoft.AspNetCore.Http.RequestDelegateGenerator.RequestDelegateGenerator\GeneratedRouteBuilderExtensions.g.cs:line 84
   at Microsoft.AspNetCore.Routing.RouteEndpointDataSource.CreateRouteEndpointBuilder(RouteEntry entry, RoutePattern groupPrefix, IReadOnlyList`1 groupConventions, IReadOnlyList`1 groupFinallyConventions)
   at Microsoft.AspNetCore.Routing.RouteEndpointDataSource.get_Endpoints()
   at Microsoft.AspNetCore.Routing.CompositeEndpointDataSource.CreateEndpointsUnsynchronized()
   at Microsoft.AspNetCore.Routing.CompositeEndpointDataSource.EnsureEndpointsInitialized()
   at Microsoft.AspNetCore.Routing.CompositeEndpointDataSource.get_Endpoints()
   at Microsoft.AspNetCore.Routing.DataSourceDependentCache`1.Initialize()
   at System.Threading.LazyInitializer.EnsureInitializedCore[T](T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory)
   at Microsoft.AspNetCore.Routing.DataSourceDependentCache`1.EnsureInitialized()
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationPolicyCache..ctor(EndpointDataSource dataSource)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware..ctor(RequestDelegate next, IAuthorizationPolicyProvider policyProvider, IServiceProvider services)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware..ctor(RequestDelegate next, IAuthorizationPolicyProvider policyProvider, IServiceProvider services, ILogger`1 logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddlewareInternal..ctor(RequestDelegate next, IServiceProvider services, IAuthorizationPolicyProvider policyProvider, ILogger`1 logger)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
   at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.ReflectionMiddlewareBinder.CreateMiddleware(RequestDelegate next)
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.Build()
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__14_1(IHostedService service, CancellationToken token)
   at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Program.<Main>$(String[] args) in D:\dev\halter73\JwtBearerAot\Program.cs:line 9
```
</details>

@eerhardt @captainsafia

### PR DESCRIPTION
**Pull request by captainsafia**
Fixes https://github.com/dotnet/aspnetcore/issues/49855.
Fixes https://github.com/dotnet/aspnetcore/issues/49849.

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt [hunk 1]
```diff
@@ -82,8 +83,11 @@ namespace Microsoft.AspNetCore.Http.Generated
...
                 var serviceProvider = options.ServiceProvider ?? options.EndpointBuilder.ApplicationServices;
                 var logOrThrowExceptionHelper = new LogOrThrowExceptionHelper(serviceProvider, options);
-                var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
-                var objectJsonTypeInfo = (JsonTypeInfo<object?>)jsonOptions.SerializerOptions.GetTypeInfo(typeof(object));
+                var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? FallbackJsonOptions;
+                var jsonSerializerOptions = jsonOptions.SerializerOptions;
+                jsonSerializerOptions.TypeInfoResolver ??= JsonTypeInfoResolver.Combine();
+                jsonSerializerOptions.MakeReadOnly();
+                var objectJsonTypeInfo = (JsonTypeInfo<object?>)jsonSerializerOptions.GetTypeInfo(typeof(object));
                 var todo_JsonTypeInfo =  (JsonTypeInfo<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo>)jsonOptions.SerializerOptions.GetTypeInfo(typeof(global::Microsoft.AspNetCore.Http.Generators.Tests.Todo));
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:91]:
```suggestion
                var todo_JsonTypeInfo =  (JsonTypeInfo<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo>)jsonSerializerOptions.GetTypeInfo(typeof(global::Microsoft.AspNetCore.Http.Generators.Tests.Todo));
```

File: src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs [hunk 2]
```diff
@@ -506,6 +506,7 @@ namespace Microsoft.AspNetCore.Http.Generated
     {{GeneratedCodeAttribute}}
     file static class GeneratedRouteBuilderExtensionsCore
     {
+        private static readonly JsonOptions FallbackJsonOptions = new();
```
Comment (eiriktsarpalis) [inline src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: You should try setting a resolver here, otherwise the singleton itself would be susceptible to races as multiple threads attempt to initialize it.

Comment (eerhardt) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: The JsonSerializerOptions and JsonTypeInfoResolver get lazy initialized in JsonOptions here:

https://github.com/dotnet/aspnetcore/blob/9427854d4f10f62a5f7f89988ea367b278ea1703/src/Http/Http.Extensions/src/JsonOptions.cs#L27-L34

when `IsReflectionEnabledByDeafult=false`, the Resolver may be null. We should fix the thread safety concerns all up (RDG, RDF, and MVC) with https://github.com/dotnet/aspnetcore/issues/49849.

Comment (eiriktsarpalis) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: It would depend on how `DefaultSerializerOptions` is defined. If it contains a resolver it would get picked up by the copy constructor.

Apropos, does the singleton need to be `JsonOptions` here. Seems it might as well be a `JsonSerializerOptions`.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt [hunk 3]
```diff
@@ -80,8 +81,11 @@ namespace Microsoft.AspNetCore.Http.Generated
                 var handler = Cast(del, global::System.String () => throw null!);
                 EndpointFilterDelegate? filteredInvocation = null;
                 var serviceProvider = options.ServiceProvider ?? options.EndpointBuilder.ApplicationServices;
-                var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
-                var objectJsonTypeInfo = (JsonTypeInfo<object?>)jsonOptions.SerializerOptions.GetTypeInfo(typeof(object));
+                var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? FallbackJsonOptions;
+                var jsonSerializerOptions = jsonOptions.SerializerOptions;
+                jsonSerializerOptions.TypeInfoResolver ??= JsonTypeInfoResolver.Combine();
```
Comment (eiriktsarpalis) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt:86]: Assuming the singleton is fully configured, couldn't this be written without any mutations as follows?
```suggestion
                var jsonSerializerOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value?.SerializerOptions;
                if (jsonSerializerOptions is null or { TypeInfoResolver: null })
                {
                     // FallbackJsonOptions configured with the empty resolver.
                     jsonSerializerOptions = FallbackJsonOptions.Value.SerializerOptions;
                }
```

File: src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: > It would depend on how DefaultSerializerOptions is defined. If it contains a resolver it would get picked up by the copy constructor.

`DefaultSerializerOptions` doesn't initialize a resolver AFAICT (it's just `new JsonSerlizerOptions(JsonDefaults.Web`) so I think being explicit here helps.

> Apropos, does the singleton need to be JsonOptions here. Seems it might as well be a JsonSerializerOptions.

True.

Comment (eerhardt) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: > It would depend on how DefaultSerializerOptions is defined.

It is part of the above linked code. I missed a couple lines.

https://github.com/dotnet/aspnetcore/blob/9427854d4f10f62a5f7f89988ea367b278ea1703/src/Http/Http.Extensions/src/JsonOptions.cs#L16-L34

> Apropos, does the singleton need to be JsonOptions here. Seems it might as well be a JsonSerializerOptions.

It should be using the `DefaultSerializerOptions` that is in JsonOptions because this is where our default JsonSerializerOptions is defined. We shouldn't be defining a new one here in generated code.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt (hunk 3 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt:86]: Note that `JsonOptions FallbackJsonOptions = new();` does NOT have an empty resolver when `JsonSerializer.IsReflectionEnabledByDefault == false`.

Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt:86]: That would be great...wouldn't we run the risk of overriding scenarios where the user set the serializer options without an explicit type info resolver but other properties set?

```
builder.Services.ConfigureHttpJsonOptions(o => o.SerializerOptions.AllowTrailingCommas = true);
```

Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt:86]: > Note that JsonOptions FallbackJsonOptions = new(); does NOT have an empty resolver when JsonSerializer.IsReflectionEnabledByDefault == false.

Correct! One implication that this has for our tests is the fact that our RDG tests are implicitly using the reflection-based type resolver for scenarios where we don't define `JsonOptions`. I realized this after trying to initialize a JsonSerializerOptions from generated code and discovering the implications of _not_ using the one defined in `JsonOptions`.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs [hunk 4]
```diff
@@ -36,7 +38,7 @@ static void ProcessParameter(EndpointParameter parameter, CodeWriter codeWriter)
                 return;
             }
             var typeName = parameter.Type.ToDisplayString(EmitterConstants.DisplayFormat);
-            codeWriter.WriteLine($"var {parameter.SymbolName}_JsonTypeInfo =  (JsonTypeInfo<{typeName}>)jsonOptions.SerializerOptions.GetTypeInfo(typeof({parameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}));");
+            codeWriter.WriteLine($"var {parameter.SymbolName}_JsonTypeInfo =  (JsonTypeInfo<{typeName}>)jsonSerializerOptions.GetTypeInfo(typeof({parameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}));");
```
Comment (eerhardt) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs:41]: (super, duper nit)

```suggestion
            codeWriter.WriteLine($"var {parameter.SymbolName}_JsonTypeInfo = (JsonTypeInfo<{typeName}>)jsonSerializerOptions.GetTypeInfo(typeof({parameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}));");
```

**Review by eerhardt — APPROVED** [review]
This LGTM. I'm slightly nervous about changing the TypeInfoResolver to be non-null so late, but I think this is the best solution to this problem.

Note that this change also fixes #49849. That should be closed when this is merged.

File: src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:509]: Couldn't it just be referencing JsonOptions.DefaultSerializerOptions directly without wrapping it in a JsonOptions then?

File: src/Mvc/Mvc.Core/test/Formatters/SystemTextJsonOutputFormatterTest.cs [hunk 5]
```diff
@@ -262,8 +262,8 @@ public void WriteResponseBodyAsync_Works_WhenTypeResolverIsNull()
         var jsonOptions = new JsonOptions();
         jsonOptions.JsonSerializerOptions.TypeInfoResolver = null;
 
-        var stjOutputFormatter = SystemTextJsonOutputFormatter.CreateFormatter(jsonOptions);
-        Assert.IsAssignableFrom<IJsonTypeInfoResolver>(stjOutputFormatter.SerializerOptions.TypeInfoResolver);
+        var exception = Assert.Throws<InvalidOperationException>(() => SystemTextJsonOutputFormatter.CreateFormatter(jsonOptions));
```
Comment (eerhardt) [inline src/Mvc/Mvc.Core/test/Formatters/SystemTextJsonOutputFormatterTest.cs:265]: (nit) the test name should probably be updated since this scenario no longer "works".

File: src/Http/Http.Extensions/src/JsonOptions.cs [hunk 6]
```diff
@@ -23,8 +23,8 @@ public class JsonOptions
 
         // The JsonSerializerOptions.GetTypeInfo method is called directly and needs a defined resolver
         // setting the default resolver (reflection-based) but the user can overwrite it directly or by modifying
-        // the TypeInfoResolverChain
-        TypeInfoResolver = JsonSerializer.IsReflectionEnabledByDefault ? CreateDefaultTypeResolver() : null
+        // the TypeInfoResolverChain. Use JsonTypeInfoResolver.Combine() to produce an empty TypeInfoResolver.
+        TypeInfoResolver = JsonSerializer.IsReflectionEnabledByDefault ? CreateDefaultTypeResolver() : JsonTypeInfoResolver.Combine()
```
Comment (eiriktsarpalis) [inline src/Http/Http.Extensions/src/JsonOptions.cs:27]: When trying to serialize with `PublishTrimmed == true` this will throw a "No metadata for type 'Foo'". Consider authoring a custom resolver that throws a desirable error message. You could also have STJ do it for you by setting the value to `JsonSerializerOptions.Default.TypeInfoResolver` (although that would require a suppresion).

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt (hunk 3 above)
Comment (halter73) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/HandlesEndpointsWithAndWithoutDiagnostics.generated.txt]: > Note that `JsonOptions FallbackJsonOptions = new();` does NOT have an empty resolver when `JsonSerializer.IsReflectionEnabledByDefault == false`.

That's true for `JsonOptions FallbackJsonOptions = new();`, but that's not necessarily true for `serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value`.

Someone could call `JsonOptions.SerializerOptions.TypeInfoResolver = null`, but I see we've added a test for this. Now that we no longer assign to `JsonOptions.SerializerOptions.TypeInfoResolver` when its null like the code originally referenced in this comment, I think we're fine.
